design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,466 @@
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: myvmi
+spec:
+  ...
+status:
+  ...
+  interfaces:
+      - infoSource: 'domain, guest-agent, multus-status'
+        podInterfaceName: custom-iface
+        interfaceName: eth0
+        ipAddress: 10.131.0.29
+        ipAddresses:
+          - 10.131.0.29
+        mac: '52:54:00:c0:35:78'
+        name: passtnet
+        queueCount: 1
+  ...
+```
+
+#### Pros
+
+- Manual configuration is not required
+- Provides backward compatibility
+- This solution has the potential to ease future changes in the name generation, as it is generated once at the
+  controller and not at each component (which may cause compatibility issues when the system is upgraded).
+
+#### Cons
+
+- Adds API
+- Exposes an implementation detail to VM owners which is probably not useful for them.
+
+### Alternative 2 - Report Primary Pod Interface Name Per VM
+
+Currently, the `VirtualMachineInstanceStatus` struct holds a slice of interface status objects:
+
+```go
+type VirtualMachineInstanceStatus struct {
+...
+Interfaces []VirtualMachineInstanceNetworkInterface `json:"interfaces,omitempty"`
+...
+}
+```
+
+A new field will be added to the `VirtualMachineInstanceStatus` struct, which will be used to report the pod's primary
+interface's name:
+
+```Go
+type VirtualMachineInstanceStatus struct {
+...
+Interfaces []VirtualMachineInstanceNetworkInterface `json:"interfaces,omitempty"`
+
+// PrimaryPodInterfaceName represents the pod primary NIC's name
+PrimaryPodInterfaceName string `json:primaryPodInterfaceName,omitempty`
+...
+}
+```
+
+The VMI controller will set this field's value from Multus' network-status annotation or fallback to `eth0`.
+
+virt-handler and virt-launcher will be adjusted to read this value, and take it into account when performing the naming scheme calculation.
+In case the field will be empty, virt-handler and virt-launcher will fall back to the existing naming scheme calculation based on the VMI.spec.
+
+
+#### API Example
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: myvmi
+spec:
+  ...
+status:
+  ...
+  primaryPodInterfaceName: custom-iface
+  interfaces:
+    - infoSource: 'domain, guest-agent, multus-status'
+      interfaceName: eth0
+      ipAddress: 10.131.0.29
+      ipAddresses:
+        - 10.131.0.29
+      mac: '52:54:00:c0:35:78'
+      name: passtnet
+      queueCount: 1
+  ...
+```
+
+#### Pros
+
+- Manual configuration is not required
+- Provides backward compatibility
+
+#### Cons
+
+- Adds API
+- Exposes an implementation detail to VM owners which is probably not useful for them.
+- This field is not relevant in case the VM is using only secondary networks
+
+### Alternative 3 - Use annotations instead of K8s API
+
+A new annotation will be added to the VirtualMachineInstance object.
+It could represent either alternative 1 or 2 in annotation form.
+
+#### Pros
+
+- Manual configuration is not required
```

Comment:
No alternative here requires manual configuration, so I would remove these from all pros. Or did I miss one?

Reply:
The KubeVirt CR alternative requires a manual configuration / operator to configure it.
I've removed this con.